### PR TITLE
test: minimal json schema only with required fields

### PR DESCRIPTION
**Pull request by hakandilek**
According to [the json schema](https://cyclonedx.org/docs/1.4/json/) only the `bomFormat`, `specVersion` and `version` fields are required. `serialNumber`  (`bom.uuid` which is RECOMMENDED but not REQUIRED), `metadata and `externalReferences` could be empty.

I've put some fixtures and tests together which only have those required fields according to the spec. The tests are unfortunately failing due to several reasons:
 - `metadata` is always assumed to be there. It could be empty or `None`. 
   - Tweaked this in the `assertEqualJsonBom` method in [tests/base.py](tests/base.py). See 0916ff38c0ddb94cde1f18182fe8b40dfc6d8c7c and
   - in several places where metadata is accessed. see 573357f2ad49aeeb5bef1a644c09c9390f6f39cd
 - `serialNumber`  (`bom.uuid`) is also assumed to exist for the tests.

I know this is an edge case which does not make any sense but one or another field which is not required should be empty.

I was not sure if I'd open an issue for this. I think a PR is better where we could start the discussion with some examples.

### CODE REVIEW

File: cyclonedx/model/bom.py [hunk 1]
```diff
@@ -315,14 +315,15 @@ def get_component_by_purl(self, purl: Optional["PackageURL"]) -> Optional[Compon
...
-    def get_urn_uuid(self) -> str:
+    def get_urn_uuid(self) -> Optional[str]:
         """
         Get the unique reference for this Bom.
 
         Returns:
             URN formatted UUID that uniquely identified this Bom instance.
         """
-        return self.__uuid.urn
+        if self.__uuid:
```
Comment (jkowalleck) [inline cyclonedx/model/bom.py:325]: more pythonic: 
```diff
-   if self.__uuid:
-     return self.__uuid.urn
+   return self.__uuid.urn if self.__uuid else None
```

File: cyclonedx/model/bom.py [hunk 2]
```diff
@@ -280,7 +280,7 @@ def metadata(self) -> BomMetaData:
         return self._metadata
 
     @metadata.setter
-    def metadata(self, metadata: BomMetaData) -> None:
+    def metadata(self, metadata: Optional[BomMetaData]) -> None:
```
Comment (jkowalleck) [inline cyclonedx/model/bom.py:283]: why? 
MEtadata is just an data-class.
If it was empty, we could just omit it from rendering/serializing when it comes to it.
Why should we have it optional?

if it was optional, then the downstream implementations would be horrible - as you always would need to check if metadata is there. Ans like you showed with the already done implementations in <https://github.com/CycloneDX/cyclonedx-python-lib/pull/327/files#diff-9c84a99c5332704dd6a90a4defadfd9df437336f07b79b91033e0a96bbd8fbf1> and others ... 
```py
if not bom.metadata:
  bom.metatada = new BomMetadata() 
bom.metadata.timestamp = now
bom.metadata.whatsoever = something
...
```

What is the point in making the models match the JSON 1:1 ? 
The JSON is just one representation of the models. just like the XML is, and the ProtoBuff is. 

---

The models are a general purpose thing with a design. 
The JSON/XML/PB are bound to the constraints of their specific systems.

**Comment by jkowalleck** [comment]
In general, I am against these proposed changes.

The models are not a representation of the JSON schema, nor the XML schema, nor the ProtoBuff representation.
The models are a general purpose thing with a certain design.
The JSON/XML/PB are bound to the constraints of their specific systems.

The models are NOT simply a 1:1 of one of the schema definitions, but are intended to be python-specific and pythonic and serve downstream implementations.
The purpose of being JSON/XML schema compatible is taken care by de-/normalizers and de-/serializers which are a decoupled topic.

Metadata is a DataClass. Having it nullable makes implementations downstream more cumbersome, as they always need to check if the `bom.metatada` exists, before accessing it.
Therefore, it was designed to be mandatory.

Same for UUID.

The `__init__` actually takes care of these facts and creates the empty nested models, if they were not provided.
Therefore, I do not see a benefit in having these optional.

**Comment by hakandilek** [comment]
I see and understand your points with the DataClass and the (de)serialization strategy. I think both are valid. 

Nevertheless, according to the spec, this is just a valid BOM (apart from the discussion of whether it's a meaningful one):

```
{
  "$schema": "http://cyclonedx.org/schema/bom-1.4.schema.json",
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "version": 1
}
```

I wanted to make sure that this _edge case_ gets tested and supported by the API.  `Optional[]` changes were just my proposal to get this done. 

I understand your concerns about that. If you still think testing and supporting the edge case makes sense, we can discuss an alternative. I'll see to revise the changes. WDYT?

**Comment by jkowalleck** [comment]
@hakandilek , if a test for the edge case of a minimal BOM was the reason, then 
please revert the changes to the models, and just leave the [`get_bom_minimal()` test you've added](https://github.com/CycloneDX/cyclonedx-python-lib/pull/327/files#diff-5c5a2b494bfc749a518e920d26f6a2808d95a389ab032375dc2566f30e540f92)
and the resulting fixtures.

**Comment by hakandilek** [comment]
> please revert the changes to the models, and just leave the [`get_bom_minimal()` test you've added](https://github.com/CycloneDX/cyclonedx-python-lib/pull/327/files#diff-5c5a2b494bfc749a518e920d26f6a2808d95a389ab032375dc2566f30e540f92) and the resulting fixtures.
> 
> Please also add the test for the XML tests, please.

sure, will do :+1:
